### ISSUE TITLE
Proxy utilisation

### ISSUE THREAD

**Issue by Ery4z** [nbd-wtf/go-nostr#75]
Im working on trying to implement an app using nostr over the tor network. Sadly I cannot find any documentation if it is possible to ask this library to communicate through a proxy (which would be the running tor instance) for the wss. Do anyone have a clue ? I tried to dive into the low level code of this library to try to understand if it would be able to set it up but im not skilled enough to do it on my own.
Otherwise if someone could help me by giving me some insight on which object structure in the library (the connection object ?) would be the most suitable to handle the proxy thing I could try to implement it. 

Thanks a lot

**Comment by fiatjaf** [comment]
By default all HTTP connections made on Go use https://pkg.go.dev/net/http#ProxyFromEnvironment and I think no library overwrites that default unless asked to.

So in this library, like many others, you can set the environment variable `HTTP_PROXY` prior to running the program and that should work.

**Comment by fiatjaf** [comment]
Please reopen if it doesn't work.

**Comment by Ery4z** [comment]
**TLDR: I get no result from the environment variable but get successful one by manually setting the proxy in the library **. 
I guess I may not have understood how to use the environment variable (even if it seems pretty basic so I don't know how im not getting it). In that case I guess it would be valuable to give a small example to anyone who would be struggling with that. 

If you think this behaviour is not normal I can propose a PR.

Thanks a lot for your time

---

I tried to make it work by running my application as this:

HTTP_PROXY="socks5://localhost:9050" ./myapp

Where "socks5://localhost:9050" point to my working container running tor.
Sadly this seems not to work as wire shark doesn't intercept any packets when sending or receiving events using the default config and exemples provided by the library.

I tried to modify the relay implementation by adding a new function 
```go
// RelayConnectProxy returns a relay object connected to url through a proxy.
// Once successfully connected, cancelling ctx has no effect.
// To close the connection, call r.Close().
func RelayConnectProxy(ctx context.Context, url string, proxy string) (*Relay, error) {
	r := &Relay{URL: NormalizeURL(url), Proxy: proxy}
	err := r.Connect(ctx)
	return r, err
}
```
adding a new Proxy field to Relay

```diff
type Relay struct {
	URL           string
	RequestHeader http.Header // e.g. for origin header

	Connection    *Connection
	subscriptions s.MapOf[string, *Subscription]

	Challenges        chan string // NIP-42 Challenges
	Notices           chan string
	ConnectionError   error
	ConnectionContext context.Context // will be canceled when the connection closes

	okCallbacks s.MapOf[string, func(bool, string)]
	mutex       sync.RWMutex

	// custom things that aren't often used
	//
	AssumeValid bool // this will skip verifying signatures for events received from this relay

+	Proxy string // if set, use this proxy for websocket connections
}
```

and by modifying Connect as this to explicitly set the proxy
```diff
func (r *Relay) Connect(ctx context.Context) error {
	connectionContext, cancel := context.WithCancel(ctx)
	r.ConnectionContext = connectionContext

	if r.URL == "" {
		cancel()
		return fmt.Errorf("invalid relay URL '%s'", r.URL)
	}

	if _, ok := ctx.Deadline(); !ok {
		// if no timeout is set, force it to 7 seconds
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, 7*time.Second)
		defer cancel()
	}

+     dialer := websocket.DefaultDialer

+	if r.Proxy != "" {
+		pUrl, err := url.Parse(r.Proxy)
+		if err != nil {
+			cancel()
+			return fmt.Errorf("error parsing proxy '%s': %w", r.Proxy, err)
+		}
+		dialer.Proxy = http.ProxyURL(pUrl)
+	}

+	socket, _, err := dialer.DialContext(ctx, r.URL, r.RequestHeader)

-      socket, _, err := web socket.DefaultDialer.DialContext(ctx, r.URL, r.RequestHeader)

	if err != nil {
		cancel()
		return fmt.Errorf("error opening websocket to '%s': %w", r.URL, err)
	}

	r.Challenges = make(chan string)
	r.Notices = make(chan string)
        ...

```

In my code I then use `RelayConnectProxy` instead of `RelayConnect`.

In that setup Wireshark intercept successfully the packets and I am able to see the bit of latency between two clients where one is using tor and the other not.

**Comment by frnandu** [comment]
I'm trying to use it behind a corporate firewall which allows only by http proxy access to outside.
I've set HTTP_PROXY and HTTPS_PROXY, but it doesn't seem to work.
Are you sure that `dialer := ws.Dialer` will use default http client with ProxyFromEnvironment?

**Comment by fiatjaf** [comment]
I am not sure, but I remember checking and concluding that yes, it would.
